### PR TITLE
Applied commits from PR #1628

### PR DESCRIPTION
**Pull request by htibosch**
Xilinx Zynq NetworkInterface v3
-----------

Description
-----------
My PR #1628 was closed because another PR #1785 came in between.

In the first commit of this PR I'm restoring the situation as it was when @gkwicker approved the PR. I hope that makes it easier to review the last few changes.

I tested all changes both in [aws_demos](https://github.com/aws/amazon-freertos/tree/master/projects/xilinx/microzed/xsdk/aws_demos) as well as in my own Zybo testing project.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c [hunk 1]
```diff
@@ -1,27 +1,27 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://aws.amazon.com/freertos
+ * http://www.FreeRTOS.org
+ */
```
Comment (htibosch) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: I am not sure about the proper style of header here

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c [hunk 2]
```diff
@@ -108,7 +108,7 @@ static BaseType_t prvGMACWaitLS( TickType_t xMaxTime );
 static void prvEMACHandlerTask( void *pvParameters );
 
 #if ( ipconfigHAS_PRINTF != 0 )
-    static void prvMonitorResources( void );
+	static void prvMonitorResources( void );
 #endif
```
Comment (htibosch) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: Replaces spaces with tab in several locations

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c [hunk 3]
```diff
@@ -326,49 +326,50 @@ BaseType_t xReturn;
...
+		{
+			/* The logging produced below may be helpful
+			 * while tuning +TCP: see how many buffers are in use. */
+			uxLastMinBufferCount = uxCurrentBufferCount;
+			FreeRTOS_printf( ( "Network buffers: %lu lowest %lu\n",
+							   uxGetNumberOfFreeNetworkBuffers(),
+							   uxCurrentBufferCount ) );
+		}
+
+		uxMinSize = xPortGetMinimumEverFreeHeapSize();
```
Comment (htibosch) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: Spaces have become tabs

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c [hunk 4]
```diff
@@ -392,11 +393,11 @@ const TickType_t ulMaxBlockTime = pdMS_TO_TICKS( 100UL );
...
 	for( ;; )
 	{
-        #if ( ipconfigHAS_PRINTF != 0 )
-            {
-                prvMonitorResources();
-            }
-        #endif /* ipconfigHAS_PRINTF != 0 ) */
+		#if ( ipconfigHAS_PRINTF != 0 )
+			{
+//				prvMonitorResources();
```
Comment (htibosch) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: Sorry, I will re-enable this again

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c [hunk 5]
```diff
@@ -86,31 +42,52 @@
...
 #include "Zynq/x_topology.h"
 #include "xstatus.h"
+
 #include "xparameters.h"
 #include "xparameters_ps.h"
 #include "xil_exception.h"
 #include "xil_mmu.h"
 
-/* Dem includes. */
 #include "uncached_memory.h"
```
Comment (htibosch) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c]: The copy of uncached_memory.{c, h} was outdated compared to the FreeRTOS+TCP release.
I cleaned it further up, mostly style changes.

File: vendors/xilinx/boards/microzed/aws_demos/config_files/FreeRTOSConfig.h [hunk 6]
```diff
@@ -165,7 +165,10 @@ extern void vLoggingPrint( const char * pcMessage );
 
 /* Map the logging task's printf to the board specific output function. */
 #include "xil_printf.h"
-#define configPRINT_STRING( X )    xil_printf( X );
+#define configPRINT_STRING( X ) \
+	xil_printf( X ); \
+	vTaskDelay( 20 )
+
```
Comment (htibosch) [inline vendors/xilinx/boards/microzed/aws_demos/config_files/FreeRTOSConfig.h]: This is a very important change for debugging: `xil_printf()` gets overflown with logging messages and 75% of the messages got lost. A short pause of 20 clock ticks makes sure all messages are sent.

File: vendors/xilinx/boards/microzed/aws_demos/config_files/FreeRTOSConfig.h [hunk 7]
```diff
@@ -165,9 +165,12 @@ extern void vLoggingPrint( const char * pcMessage );
 
 /* Map the logging task's printf to the board specific output function. */
 #include "xil_printf.h"
+
+#define configPAUSE_AFTER_XIL_PRINTF_MSEC	20
+
 #define configPRINT_STRING( X ) \
 	xil_printf( X ); \
-	vTaskDelay( 20 )
+	vTaskDelay( pdMS_TO_TICKS( configPAUSE_AFTER_XIL_PRINTF_MSEC ) )
```
Comment (htibosch) [inline vendors/xilinx/boards/microzed/aws_demos/config_files/FreeRTOSConfig.h]: Replaced the magic number 20 with a macro with a clear name

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c (hunk 1 above)
Comment (gkwicker) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: This looks correct.

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c [hunk 8]
```diff
@@ -206,7 +221,23 @@ const TickType_t xWaitLinkDelay = pdMS_TO_TICKS( 7000UL ), xWaitRelinkDelay = pd
...
 {
-	if( ( ulPHYLinkStatus & BMSR_LINK_STATUS ) != 0 )
+	#if( ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM != 0 )
+	{
+	ProtocolPacket_t *pxPacket;
+
+		/* If the peripheral must calculate the checksum, it wants
+		the protocol checksum to have a value of zero. */
+		pxPacket = ( ProtocolPacket_t * ) ( pxBuffer->pucEthernetBuffer );
+		if( pxPacket->xICMPPacket.xIPHeader.ucProtocol == ipPROTOCOL_ICMP )
```
Comment (gkwicker) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c]: Let's future-proof ourselves here and make this comparison explicit:

```c
if(( pxPacket->xICMPPacket.xIPHeader.ucProtocol != ipPROTOCOL_UDP ) &&
   ( pxPacket->xICMPPacket.xIPHeader.ucProtocol != ipPROTOCOL_TCP ))
{
.
.
.
}
```

Someone might want to add [IGMP](https://en.wikipedia.org/wiki/Internet_Group_Management_Protocol) support someday, you never know.

File: libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c [hunk 9]
```diff
@@ -159,29 +137,29 @@ uint32_t ulSkipSize;
...
-		vLoggingPrintf( "vInitialiseUncachedMemory: Can not allocate uncached memory\n" );
+		FreeRTOS_printf( ( "vInitialiseUncachedMemory: Can not allocate uncached memory\n" ) );
 	}
 	else
 	{
 		/* Some objects want to be stored in uncached memory. Hence the 1 MB
 		address range that starts after "_end" is made uncached by setting
 		appropriate attributes in the translation table. */
-		Xil_SetTlbAttributes( ( uint32_t )pucStartOfMemory, 0xc02 ); /* addr, attr */
+		/* FIXME claudio rossi. Modified to prevent data abort exception (misaligned access)
```
Comment (gkwicker) [inline libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c]: Is this `FIXME` still accurate?  If so, can we eliminate it in this revision?

Comment (htibosch) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c]: > Is this FIXME still accurate?

In fact, nothing has to be fixed. A developer Claudio Rossi proposed to change the memory attribute from `0xC02` to `0x1C02`:

`#define uncMEMORY_ATTRIBUTE       0x1C02`

I do not know what is the meaning of the extra bit `0x1000`, but it works equally well.

I don't mind removing the text.

Comment (htibosch) [reply libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/uncached_memory.c]: > I do not know what is the meaning of the extra bit `0x1000`

Looking at the earlier comment: it allows unaligned access in the specified memory range.

> @gkwicker requested changes 

Are those changes now satisfactory?

Anyone else wants to review this PR? 
@yuhui-zheng maybe?

Note, this PR received the wrong title. It should have bee something like:
**FreeRTOS+TCP : Revision of Xilinx Network Interface Version 2**

**Comment by leegeth** [comment]
There are build errors for xilinx projects for both demos and tests

afr_pr_build_win/libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.c:91:3: error: #error Please define both 'ipconfigZERO_COPY_RX_DRIVER' and 'ipconfigZERO_COPY_TX_DRIVER' as 1
  #error Please define both 'ipconfigZERO_COPY_RX_DRIVER' and 'ipconfigZERO_COPY_TX_DRIVER' as 1
   ^~~~~
make: *** [libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/Zynq/NetworkInterface.o] Error 1

Are there any associated config updates of ipconfigZERO_COPY_RX_DRIVER and ipconfigZERO_COPY_TX_DRIVER for xilinx missed?

**Comment by htibosch** [comment]
> Are there any associated config updates of ipconfigZERO_COPY_RX_DRIVER
> and ipconfigZERO_COPY_TX_DRIVER for xilinx missed?

Sorry, yes there are. I forgot commit both copies of `FreeRTOSIPConfig.h`.

Thanks, Hein

**Pull request merged**